### PR TITLE
[ONNX] Enable scripting tests and update jit passes

### PR DESCRIPTION
**Pull request by neginraoof**
This PR initiates the process of updating the torchsciprt backend interface used by ONNX exporter.

- Replace jit lower graph pass by freeze module pass

- Enable ScriptModule tests for ONNX operator tests (ORT backend) and model tests by default.

- Replace jit remove_inplace_ops pass with remove_mutation and consolidation all passes for handling inplace ops.

### CODE REVIEW

File: torch/onnx/utils.py
Comment (BowenBao) [inline torch/onnx/utils.py:337]: `example_outputs` seems not needed in this method.

**Review by BowenBao — COMMENTED** [review]
Test failures are related (removal of parameter `propagate`). Please add in the description or comment about context of this change as well.

File: torch/onnx/utils.py
Comment (neginraoof) [reply torch/onnx/utils.py:337]: How about for assertions?

Comment (BowenBao) [reply torch/onnx/utils.py:337]: can probably move assertions outside this method? seems only needed later

File: torch/onnx/utils.py
Comment (spandantiwari) [inline torch/onnx/utils.py:337]: Can we keep the original name with `_` for `_retain_param_name`?

File: torch/onnx/utils.py
Comment (BowenBao) [inline torch/onnx/utils.py:495]: argument `training` is missing after the change. Is this intended?

Comment (neginraoof) [reply torch/onnx/utils.py]: Thanks for the catch! No, this was just a bad merge.

**Review by BowenBao — DISMISSED** [review]
LGTM

**Review by bzinodev — APPROVED** [review]

**Comment by neginraoof** [comment]
I'm repurposing this PR and including more changes.

**Comment by mruberry** [comment]
It looks likes this broke caffe2_onnx_ort1_py3_6_clang7_ubuntu16_04_test and caffe2_onnx_ort2_py3_6_clang7_ubuntu16_04_test on master. Relevant log snippets:

```
Aug 04 02:44:59 ==================================== ERRORS ====================================
Aug 04 02:44:59 _________ ERROR collecting test/onnx/test_pytorch_onnx_onnxruntime.py __________
Aug 04 02:44:59 test/onnx/test_pytorch_onnx_onnxruntime.py:103: in <module>
Aug 04 02:44:59     class TestONNXRuntime(unittest.TestCase):
Aug 04 02:44:59 test/onnx/test_pytorch_onnx_onnxruntime.py:2700: in TestONNXRuntime
Aug 04 02:44:59     @enableScriptTest()
Aug 04 02:44:59 E   NameError: name 'enableScriptTest' is not defined
Aug 04 02:44:59 _________ ERROR collecting test/onnx/test_pytorch_onnx_onnxruntime.py __________
Aug 04 02:44:59 test/onnx/test_pytorch_onnx_onnxruntime.py:103: in <module>
Aug 04 02:44:59     class TestONNXRuntime(unittest.TestCase):
Aug 04 02:44:59 test/onnx/test_pytorch_onnx_onnxruntime.py:2700: in TestONNXRuntime
Aug 04 02:44:59     @enableScriptTest()
Aug 04 02:44:59 E   NameError: name 'enableScriptTest' is not defined
Aug 04 02:44:59 _________ ERROR collecting test/onnx/test_pytorch_onnx_onnxruntime.py __________
Aug 04 02:44:59 test/onnx/test_pytorch_onnx_onnxruntime.py:103: in <module>
Aug 04 02:44:59     class TestONNXRuntime(unittest.TestCase):
Aug 04 02:44:59 test/onnx/test_pytorch_onnx_onnxruntime.py:2700: in TestONNXRuntime
Aug 04 02:44:59     @enableScriptTest()
Aug 04 02:44:59 E   NameError: name 'enableScriptTest' is not defined
```

```
Aug 04 02:44:59 -- Docs: https://docs.pytest.org/en/latest/warnings.html
Aug 04 02:44:59 =========================== short test summary info ============================
Aug 04 02:44:59 ERROR test/onnx/test_pytorch_onnx_onnxruntime.py - NameError: name 'enableScr...
Aug 04 02:44:59 ERROR test/onnx/test_pytorch_onnx_onnxruntime.py - NameError: name 'enableScr...
Aug 04 02:44:59 ERROR test/onnx/test_pytorch_onnx_onnxruntime.py - NameError: name 'enableScr...
Aug 04 02:44:59 ======================== 12 warnings, 3 errors in 1.34s ========================
Aug 04 02:44:59 ERROR: not found: /var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset7
Aug 04 02:44:59 (no name '/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset7' in any of [<Module test/onnx/test_pytorch_onnx_onnxruntime.py>])
Aug 04 02:44:59 
Aug 04 02:44:59 ERROR: not found: /var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset8
Aug 04 02:44:59 (no name '/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset8' in any of [<Module test/onnx/test_pytorch_onnx_onnxruntime.py>])
Aug 04 02:44:59 
Aug 04 02:44:59 ERROR: not found: /var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime
Aug 04 02:44:59 (no name '/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime' in any of [<Module test/onnx/test_pytorch_onnx_onnxruntime.py>])
Aug 04 02:44:59 
```

These tests didn't fail on the PR, but the PR tests are several days old. Unlanding to unblock CI.

**Comment by spandantiwari** [comment]
@neginraoof - Can we rerun and pass all the tests so that we can merge this PR again?

**Comment by neginraoof** [comment]
Thanks. This PR needs more review as I had repurposed it and had added more changes. It needs another round of review from team. It's okay to pull this back for now.

File: torch/onnx/utils.py
Comment (BowenBao) [inline torch/onnx/utils.py:348]: The preprocess pass was designed to be applied on both tracing and scripting, and at later stage of the conversion (just before converting all nodes to onnx namespace). On the other hand, I see the necessity of putting this pass before remove_mutation. Maybe it should be put into a separate pass, one that focus on remove mutation for onnx. What do you think?

Comment (neginraoof) [reply torch/onnx/utils.py:348]: So my idea is to eventually move these two passes to the main code-path and replace remove_inplace_ops pass.
I was thinking we can run the preprocess pass twice for now.

File: torch/csrc/jit/passes/lower_tuples.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/lower_tuples.cpp:24]: What's the context for this change? Looks similar to https://github.com/pytorch/pytorch/pull/41548.

File: torch/onnx/utils.py
Comment (BowenBao) [inline torch/onnx/utils.py:507]: Maybe it's good to add comments explaining the new argument.

File: torch/onnx/utils.py
Comment (neginraoof) [reply torch/onnx/utils.py]: So the an exception msg with string.format appears in the graph as RaiseException node with a tuple which contains aten::format. So I'm adding this op type to the allowed list to enable lower_tuple pass.
This unblocks scripting a number of tests on pytorch_onnx_onnxruntime.

File: torch/csrc/jit/passes/lower_tuples.cpp
Comment (neginraoof) [reply torch/csrc/jit/passes/lower_tuples.cpp:24]: So an exception msg from torch.nn.functional with string.format appears in the graph as prim::RaiseException node with a tuple which contains aten::format. So I'm adding this op type to the allowed list to enable lower_tuple pass.
This unblocks scripting a number of tests on pytorch_onnx_onnxruntime.

File: torch/onnx/utils.py
Comment (BowenBao) [inline torch/onnx/utils.py:503]: nit
```suggestion
# Once these jit APIs are fully tested, they will become part of production code-path by
```

File: torch/onnx/utils.py
Comment (BowenBao) [inline torch/onnx/utils.py:502]:
```suggestion
# The purpose of this flag is to enable the new API temporarily for testing purposes.
```

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:451]: Maybe need to copymetadata for newNode->output() as well?

File: torch/onnx/utils.py
Comment (BowenBao) [inline torch/onnx/utils.py:137]: looks like the new pass `_jit_pass_onnx_remove_inplace_ops_for_onnx` is not being used anywhere

File: torch/csrc/jit/passes/remove_mutation.cpp
Comment (eellison) [inline torch/csrc/jit/passes/remove_mutation.cpp:267]: This is going to construct an alias DB on each invocation, which will be super slow. As such I think it's better not to expose `isInplaceOpVariant` as a public API except as a method on the class, and in whatever pass relies on this have it instantiate a single MutationRemover and query that one instance `inplaceOpVariant`. 

In the future it would be nice to decouple the alias schema querying from instantiating the alias db, but that's a good amount of work and probably not needed for this PR

**Review by eellison — COMMENTED** [review]
one suggestion would be to create a MutationRemover instance and re-query it, instead of reconstructing on each query

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp]: Thanks for mentioning this, actually only the output metadata had to be copied.

File: torch/csrc/jit/passes/remove_mutation.cpp
Comment (neginraoof) [reply torch/csrc/jit/passes/remove_mutation.cpp]: Thanks @eellison 
So for now we can use the MutationRemover aliasDb. I just made the change.

**Review by BowenBao — APPROVED** [review]
LGTM, cc @eellison for review on the remove mutation part.

File: torch/csrc/jit/passes/remove_mutation.h
Comment (eellison) [inline torch/csrc/jit/passes/remove_mutation.h:12]: Can you move some of the implementation out of the header, or use an impl pattern? Having the whole struct in the .h and then four lines of codes in the cpp file will lead to bad bulid times when I'm working on this file. 

Another, maybe easiest alternative is to add a `inplaceOpVariant` api that takes in an AliasDb and a node.
```
void InplaceOpVariant(Node *n, const AliasDb& aliasDb)
```

and then have mutation remover / your pass both invoke that

**Review by eellison — APPROVED** [review]
remove_mutation.h changes lgtm

**Review by bzinodev — APPROVED** [review]

**Comment by neginraoof** [comment]
@bzinodev Would you please take a look and merge this PR if possible? This is important for the first phase of using new scripting APIs.

**Pull request closed without merging**
